### PR TITLE
feat: deprecate all functions & types exported from `magicExports`

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
I wanted to re-submit #2735 again, as:

- The first release supporting separate packages was `1.3.3`, which was released on 23 Mar 2022
  https://github.com/remix-run/remix/releases/tag/v1.3.3
- The `replace-remix-imports` migration was added in `1.4.0`, which was released on 14 Apr 2022
  https://github.com/remix-run/remix/releases/tag/v1.4.0
- We now export functions/types that aren't available anymore through the `remix` package, as they're not included in the `magicExports` anymore (see both `@remix-run/cloudflare` & `@remix-run/node`) since `1.5.0`
- We now have a runtime package (`@remix-run/deno`) that doesn't have *any* magic exports *at all* since `1.5.0`
https://github.com/remix-run/remix/releases/tag/v1.5.0
- Our ESLint config is warning with a deprecated message when importing from the `remix` package since `1.6.0`
https://github.com/remix-run/remix/releases/tag/v1.6.0

---

Original PR description:

> As a follow-up of #2731, this will show deprecation warnings in the IDE + in console whenever people aren't using the ESLint config.
> 
> Inspired by @pcattori's remark (https://github.com/remix-run/remix/pull/2542#discussion_r838804227) on @mjackson's PR to rename `createCloudflareKVSessionStorage` to `createWorkersKVSessionStorage` (#2542). 
> 
> Since each package is on it's own, I had to copy/paste the `warn` & `getDeprecatedMessage` functions over to each package.
> 
> If this somehow could be simplified (as this is doing the same thing over & over again), I'd be happy to do so.

### CODE REVIEW

**Review by kentcdodds — APPROVED** [review]
I want @mjackson or @ryanflorence to approve this before merging.

**Comment by MichaelDeBoey** [comment]
Our ESLint config is warning with a deprecated message when importing from the `remix` package since `1.6.0`
https://github.com/remix-run/remix/releases/tag/v1.6.0

**Comment by MichaelDeBoey** [comment]
This will need a re-write now that #3543 is merged 🙈

**Comment by chaance** [comment]
@MichaelDeBoey I've pinged the bosses on this one again and am hoping to have a resolution this week. If approved, do you mind adding a changeset now that we have that work merged?

**Comment by mjackson** [comment]
I think now is a great time to officially deprecate using magic exports 👍 

Do you know why the tests are failing in the diff view @MichaelDeBoey? Is that something to do with this PR? Or just something that has been happening for a while now? I'd like to clean that up if we can. If the test failures are unrelated to this PR, it's a huge distraction.

Can you also please say more about the approach you're taking here? It looks like you're generating a bunch of code in the build output. Is that right? I think I'd prefer to just write out the code (it's repetitive, I know) instead of introduce more complexity in our build.

**Comment by MichaelDeBoey** [comment]
> Do you know why the tests are failing in the diff view?

If you're talking about the failing tests underneath `Unchanged files with check annotations`, this has been happening for a while now.

> Can you also please say more about the approach you're taking here? It looks like you're generating a bunch of code in the build output. Is that right? I think I'd prefer to just write out the code (it's repetitive, I know) instead of introduce more complexity in our build.

In #3543, @chaance removed all manual `magicExports` files & replaced it by generating them.
The generation script is taking the defined functions/types from `rollup.config.js`'s `getMagicExports` & exports them.

My changes are building upon that script, by wrapping all functions in a `warn` function that's `console.warn`ing a deprecation message before calling the actual function.

My changes also add an `/** @deprecated **/` typehint to all functions/types & points them towards using the actual packages instead of using `'remix'` package.

**Comment by chaance** [comment]
@mjackson We moved to code generation here because having the code in the workspace was causing TypeScript issues when we did our Rollup build overhaul a while back. It was always intended to be temporary as we planned on removing the exports altogether, but we went this route to move a little quicker, as that refactor was blocking other work if I recall.

File: package.json
Comment (chaance) [inline package.json:90]: I think we can remove this.

```suggestion
```

File: rollup.utils.js
Comment (chaance) [inline rollup.utils.js:199]: We'll merge this before the next minor release.
```suggestion
          "const getDeprecatedMessage = (symbol, packageName) => `All \\`remix\\` exports are considered deprecated as of v1.8.0. Please import \\`${symbol}\\` from \\`${packageName}\\` instead. You can run \\`remix migrate --migration replace-remix-imports\\` to automatically migrate your code.`;\n" +
```

File: .changeset/khaki-meals-prove.md
Comment (chaance) [inline .changeset/khaki-meals-prove.md:9]: Note: we don't need to use conventional commit formatting for these. They should generally read as plain sentences, and they'll be grouped by type in the release notes automatically by Changesets.

```suggestion
Importing functions and types from the `remix` package is deprecated, and all exported modules will be removed in the next major release. For more details, [see the release notes for 1.4.0](https://github.com/remix-run/remix/releases/tag/v1.4.0) where these changes were first announced.
```

**Review by chaance — CHANGES_REQUESTED** [review]
These look good overall. Can you merge the latest from `dev` one last time? Once that's taken care of I'll cut an experimental release to manually check the dev warnings, and if everything looks good we'll get this in for the next minor bump.

Thanks so much for seeing this through! 🙏

File: rollup.utils.js
Comment (MichaelDeBoey) [reply rollup.utils.js:199]: Since we use `v1.3.3` in our ESLint warnings as well, I kept it here too

https://github.com/remix-run/remix/blob/983836112ac7400e53cb6f35b2cbf37035d716f2/packages/remix-eslint-config/rules/core.js#L16-L17

**Comment by chaance** [comment]
@MichaelDeBoey Just FYI, the only reason I haven't merged this yet is because I want to time it with an upcoming minor release. Once @brophdawg11 and @jacob-ebey finalize their work integrating React Router 6.4 I'll merge this and ship them together.

**Comment by chaance** [comment]
@MichaelDeBoey Ready to merge this if you're able to bring the branch up-to-date 😄

**Review by chaance — APPROVED** [review]

**Pull request merged**
